EVENTS.md

```diff
@@ -0,0 +1,416 @@
+
+## Event Details for wasmd
+
+Beyond the basic Event system and emitted events, we must handle more advanced cases in `x/wasm`
+and thus add some more logic to the event processing. Remember that CosmWasm contracts dispatch other
+messages themselves, so far from the flattened event structure, or even a list of list (separated
+by message index in the tx), we actually have a tree of messages, each with their own events. And we must
+flatten that in a meaningful way to feed it into the event system.
+
+Furthermore, with the sub-message reply handlers, we end up with eg. "Contract A execute", "Contract B execute",
+"Contract A reply". If we return all events by all of these, we may end up with many repeated event types and
+a confusing results. However, we may use the standard "message" events to separate the sub-messages as it marks
+where the next one starts. With careful analysis of the "sender" field on these "message" markers, we may be able
+to reconstruct much of the tree execution path. We should ensure all this information is exposed in the most
+consistent way possible.
+
+### Combining Events from Sub-Messages
+
+Each time a contract is executed, it not only returns the `message` event from its call, the `execute` event for the
+contact and the `wasm` event with any custom fields from the contract itself. It will also return the same set of information
+for all messages that it returned, which were later dispatched.
+
+Until 0.18.0, we stripped out the `message` events that were returned when dispatching `Response.messages`. However,
+they contain useful information to trace the topography of the call, especially when we call into native modules,
+and it makese sense to maintain the full information.
+
+Note that these are all appended in order called. So if I execute a contract, which returns two messages, one to instantiate a new
+contract and the other to set the withdrawl address, while also using `ReplyOnSuccess` for the instantiation (to get the
+address), it will emit a series of events that looks something like this:
+
+```go
+/// original execution (top-level message)
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("module", "wasm"),
+    sdk.NewAttribute("sender", msg.Sender),  
+),
+sdk.NewEvent(
+    "execute",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("result", hex.EncodeToString(data)),
+),
+sdk.NewEvent(
+    "wasm",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("custom", "from contract"),
+),
+
+// instantiating contract (returned message)
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("module", "wasm"),
+    sdk.NewAttribute("sender", contractAddr.String()),  
+),
+sdk.NewEvent(
+    "instantiate",
+    sdk.NewAttribute("code_id", fmt.Sprintf("%d", msg.CodeID)),
+    sdk.NewAttribute("_contract_addr", newContract.String()),
+    sdk.NewAttribute("result", hex.EncodeToString(initData)),
+)
+sdk.NewEvent(
+    "wasm",
```

Comment:
Looking at this event with some distance now, I don't see how it adds value. Before the contract address was not available but it is now by the wasm operation event. Would be nice to drop this functionality and let people emit custom events when needed. That would be easier to explain in dev doc and no "empty" event would be emitted by default.

Reply:
So, if I return non-empty `attributes`, then it will get this `wasm` event (tagged).
If I return empty attributes, it will not get this event, right?

The one issue I want to bring up (which came up from Secret network some time ago) is they want to be able to search for all actions on a contract easily. We can now search for `execute._contract_address="foo"`. However, maybe we also want to include `instantiate._contract_address="foo"` and `migrate._contract_address="foo"`. I guess all data is there.

So, yeah, if we say don't emit `wasm` event on empty attributes, I am fine with that

Reply:
I would go one step further and drop the whole event. Why stick with a `wasm` event when we have custom events now for a contract? I don't see the use case for a client anymore. But there is complexity in wasmd to handle the case and complexity in wasm contracts for people to learn and understand the framework. Both sides would benefit from a clean design that does not support 2 ways of storing data in events.

Reply:
90% of contracts just want to emit attributes.
Only a few emit multiple events.
We didn't want to make it that much more complex for all of them, which is why we didn't remove the attributes when we introduced events. 

But you have a point. And with the new builder API from @uint we could even let people just `add_attribute` still and just expose them as a `wasm-default` event (so we keep the same cosmwasm-std api, but change the wasm interface between contract and runtime).

That is a larger discussion, and 0.16 is already cut. I feel it might be too late to make this change, but we could consider it for 1.0.0-beta.

That said, I would document both paths for now as we will support both in 0.18. If we change that for 1.0.0-beta, we can update the doc as well.